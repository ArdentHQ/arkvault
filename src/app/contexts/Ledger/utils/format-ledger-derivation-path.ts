import { LedgerDerivationScheme } from "@/app/contexts/Ledger/Ledger.contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
