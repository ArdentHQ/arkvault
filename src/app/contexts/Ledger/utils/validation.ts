import semver from "semver";
import { Coins } from "@payvo/sdk";
import { minVersionList } from "@/app/contexts/Ledger/Ledger.contracts";

export const hasRequiredAppVersion = async (coin: Coins.Coin) => {
	const coinId = coin.network().coin();

	if (minVersionList[coinId]) {
		const currentVersion = await coin.ledger().getVersion();

		if (semver.lt(currentVersion, minVersionList[coinId])) {
			return false;
		}
	}

	return true;
};
