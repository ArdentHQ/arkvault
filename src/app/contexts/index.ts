export * from "./Configuration";
export * from "./Environment";
export * from "./Ledger";
export * from "./Navigation";
export * from "./Steps";
