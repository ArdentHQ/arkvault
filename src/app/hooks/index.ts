export * from "./click-outside";
export * from "./debounce";
export * from "./env";
export * from "./network-status";
export * from "./use-accent-color";
export * from "./use-clipboard";
export * from "./use-deeplink";
export * from "./use-document-title";
export * from "./use-fees";
export * from "./use-ledger-model-status";
export * from "./use-locale-currency";
export * from "./use-network-options";
export * from "./use-password-validation";
export * from "./use-previous";
export * from "./use-profile-synchronizer";
export * from "./use-query-parameters";
export * from "./use-random-number";
export * from "./use-reload-path";
export * from "./use-scroll";
export * from "./use-synchronizer";
export * from "./use-theme";
export * from "./use-validation";
export * from "./use-wallet-alias";
export * from "./use-breakpoint";
export * from "./use-accordion";
export * from "./use-server-health-status";
