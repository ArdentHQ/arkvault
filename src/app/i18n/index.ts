import i18n from "i18next";
import { initReactI18next } from "react-i18next";

// Helpers
import { buildTranslations } from "./helpers";

const defaultNS = "translation";

const resources = {
	en: {
		translation: buildTranslations(),
	},
};

i18n.use(initReactI18next).init({
	defaultNS,
	lng: "en",
	ns: [defaultNS],
	resources,
});

export { defaultNS, resources };

export { default as i18n } from "i18next";
