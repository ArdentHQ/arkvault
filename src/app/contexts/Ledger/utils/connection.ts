import { Coins } from "@ardenthq/sdk";
import retry, { AbortError, Options } from "p-retry";

import { formatLedgerDerivationPath } from "./format-ledger-derivation-path";
import { hasRequiredAppVersion } from "./validation";

export const accessLedgerDevice = async (coin: Coins.Coin) => {
	try {
		await coin.__construct();
		await coin.ledger().connect();
	} catch (error) {
		// If the device is open, continue normally.
		// Can be triggered when the user retries ledger connection.
		if (error.message !== "The device is already open.") {
			throw error;
		}
	}
};

export const accessLedgerApp = async ({ coin }: { coin: Coins.Coin }) => {
	await accessLedgerDevice(coin);

	if (!(await hasRequiredAppVersion(coin))) {
		throw new Error("VERSION_ERROR");
	}

	// Ensure that the app is accessible.
	await coin.ledger().getPublicKey(
		formatLedgerDerivationPath({
			coinType: coin.config().get<number>("network.constants.slip44"),
		}),
	);
};

export const persistLedgerConnection = async ({
	coin,
	options,
	hasRequestedAbort,
}: {
	coin: Coins.Coin;
	options: Options;
	hasRequestedAbort: () => boolean;
}) => {
	const retryAccess: any = async (attempts: number) => {
		if (hasRequestedAbort() && attempts > 1) {
			throw new AbortError("CONNECTION_ERROR");
		}

		try {
			await accessLedgerApp({ coin });
		} catch (error) {
			// Abort on version error or continue retrying access.
			if (error.message === "VERSION_ERROR") {
				throw new AbortError("VERSION_ERROR");
			}

			throw error;
		}
	};

	await retry(retryAccess, options);
};
