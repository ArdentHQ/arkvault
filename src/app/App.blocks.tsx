import { Global, css } from "@emotion/react";
import React, { useCallback, useLayoutEffect, useRef, useState } from "react";
import { HashRouter, useHistory } from "react-router-dom";
import { useTranslation } from "react-i18next";
import { useErrorHandler } from "react-error-boundary";
import { ToastContainer } from "react-toastify";
import { GlobalStyles as BaseStyles } from "twin.macro";

import { ConfirmationModal } from "@/app/components/ConfirmationModal";
import { useEnvironmentContext, useNavigationContext } from "@/app/contexts";
import { useAccentColor, useNetworkStatus, useProfileSynchronizer, useTheme } from "@/app/hooks";
import { toasts } from "@/app/services";
import { SyncErrorMessage } from "@/app/components/ProfileSyncStatusMessage";
import { bootEnvironmentWithProfileFixtures, isE2E, isUnit } from "@/utils/test-helpers";
import { Offline } from "@/domains/error/pages";
import { middlewares, RouterView, routes } from "@/router";
import { PageSkeleton } from "@/app/components/PageSkeleton";
import { ProfilePageSkeleton } from "@/app/components/PageSkeleton/ProfilePageSkeleton";
import { InstallPWA } from "@/domains/dashboard/components/InstallPWA";

const AppRouter: React.FC = ({ children }) => {
	const [isOpen, setIsOpen] = useState(false);

	const confirmationFunctionReference = useRef<(allowNavigate: boolean) => void>();

	const onCancel = () => {
		confirmationFunctionReference.current?.(false);
		setIsOpen(false);
	};

	const onConfirm = () => {
		confirmationFunctionReference.current?.(true);
		setIsOpen(false);
	};

	const getUserConfirmation = useCallback((_, callback) => {
		confirmationFunctionReference.current = callback;
		setIsOpen(true);
	}, []);

	return (
		<React.Suspense fallback={<PageSkeleton />}>
			<HashRouter getUserConfirmation={getUserConfirmation}>
				{children}
				<ConfirmationModal isOpen={isOpen} onCancel={onCancel} onConfirm={onConfirm} />
			</HashRouter>
		</React.Suspense>
	);
};

const customStyles = css`
	h1 {
		margin-bottom: 0.5rem;
		font-size: 2.25rem;
		line-height: 2.5rem;
		font-weight: 700;
	}
	h2 {
		margin-bottom: 0.5rem;
		font-size: 1.875rem;
		line-height: 2.25rem;
		font-weight: 700;
	}
	h3 {
		margin-bottom: 0.5rem;
		font-size: 1.5rem;
		line-height: 2rem;
		font-weight: 600;
	}
`;

const GlobalStyles: React.VFC = () => (
	<>
		<BaseStyles />
		<Global styles={customStyles} />
	</>
);

const Main: React.VFC = () => {
	const { env, persist, isEnvironmentBooted, setIsEnvironmentBooted } = useEnvironmentContext();
	const isOnline = useNetworkStatus();
	const history = useHistory();

	const { resetAccentColor } = useAccentColor();
	const { resetTheme } = useTheme();

	const { setShowMobileNavigation } = useNavigationContext();

	const { t } = useTranslation();

	useProfileSynchronizer({
		onProfileRestoreError: () =>
			history.push({
				pathname: "/",
				state: {
					from:
						history.location.pathname + history.location.search ??
						/* istanbul ignore next -- @preserve */ "",
				},
			}),
		onProfileSignOut: () => {
			resetTheme();
			resetAccentColor();
			toasts.dismiss();

			setShowMobileNavigation(false);
		},
		onProfileSyncComplete: async () => {
			await toasts.dismiss();
			toasts.success(t("COMMON.PROFILE_SYNC_COMPLETED"));
		},
		onProfileSyncError: async (failedNetworkNames, retryProfileSync) => {
			await toasts.dismiss();

			toasts.warning(
				<SyncErrorMessage
					failedNetworkNames={failedNetworkNames}
					onRetry={async () => {
						await toasts.dismiss();
						retryProfileSync();
					}}
				/>,
			);
		},
		onProfileSyncStart: () => {
			toasts.warning(t("COMMON.PROFILE_SYNC_STARTED"), { autoClose: false });
		},
		onProfileUpdated: () => {
			history.replace("/");
		},
	});

	const handleError = useErrorHandler();

	useLayoutEffect(() => {
		const boot = async () => {
			try {
				/* istanbul ignore next -- @preserve */
				if (isE2E() || isUnit()) {
					await bootEnvironmentWithProfileFixtures({ env, shouldRestoreDefaultProfile: isUnit() });
					await persist();

					setIsEnvironmentBooted(true);
					return;
				}

				/* istanbul ignore next -- @preserve */
				await env.verify();

				/* istanbul ignore next -- @preserve */
				await env.boot();

				setIsEnvironmentBooted(true);
			} catch (error) {
				handleError(error);
			}
		};

		boot();
	}, [env, handleError]);

	const Skeleton = history.location.pathname.startsWith("/profiles") ? ProfilePageSkeleton : PageSkeleton;

	const renderContent = () => {
		if (!isOnline) {
			return <Offline />;
		}

		/* istanbul ignore else -- @preserve */
		if (isEnvironmentBooted) {
			return <RouterView routes={routes} middlewares={middlewares} />;
		}

		return <Skeleton />;
	};

	return (
		<main data-testid="Main">
			<InstallPWA />

			<ToastContainer closeOnClick={false} newestOnTop />

			{renderContent()}
		</main>
	);
};

export { AppRouter, GlobalStyles, Main };
