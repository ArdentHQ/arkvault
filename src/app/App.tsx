import "focus-visible";

import React, { useRef } from "react";
import { ErrorBoundary } from "react-error-boundary";
import { I18nextProvider } from "react-i18next";
import { CacheProvider } from "@emotion/react";

import createCache from "@emotion/cache";
import { ConfigurationProvider, EnvironmentProvider, LedgerProvider, NavigationProvider } from "./contexts";
import { i18n as index18n } from "./i18n";
import { ZendeskProvider } from "./contexts/Zendesk";
import { ExchangeProvider } from "@/domains/exchange/contexts/Exchange";
import { AppRouter, GlobalStyles, Main } from "@/app/App.blocks";
import { initializeEnvironment } from "@/utils/environment";
import { ApplicationError } from "@/domains/error/pages";

export const App: React.VFC = () => {
	/**
	 * Ensure that the Environment object will not be recreated when the state changes,
	 * as the data is stored in memory by the `DataRepository`.
	 */
	const environment = useRef(initializeEnvironment());

	const cache = createCache({ key: "emotion-cache" });
	cache.compat = true;

	return (
		<ZendeskProvider>
			<I18nextProvider i18n={index18n}>
				<EnvironmentProvider env={environment.current}>
					<ConfigurationProvider defaultConfiguration={{ profileIsSyncingExchangeRates: true }}>
						<NavigationProvider>
							<ExchangeProvider>
								<ErrorBoundary FallbackComponent={ApplicationError}>
									<LedgerProvider>
										<CacheProvider value={cache}>
											<AppRouter>
												<GlobalStyles />
												<Main />
											</AppRouter>
										</CacheProvider>
									</LedgerProvider>
								</ErrorBoundary>
							</ExchangeProvider>
						</NavigationProvider>
					</ConfigurationProvider>
				</EnvironmentProvider>
			</I18nextProvider>
		</ZendeskProvider>
	);
};
