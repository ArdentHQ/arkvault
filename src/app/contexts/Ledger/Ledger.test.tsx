/* eslint-disable @typescript-eslint/require-await */
import React from "react";
import { EnvironmentProvider, LedgerProvider } from "@/app/contexts";
import { env, render, screen } from "@/utils/testing-library";

describe("Ledger Provider", () => {
	it("should render ledger provider", async () => {
		process.env.REACT_APP_IS_UNIT = "1";

		const { container, asFragment } = render(
			<EnvironmentProvider env={env}>
				<LedgerProvider>
					<div data-testid="content" />
				</LedgerProvider>
			</EnvironmentProvider>,
		);

		expect(screen.getByTestId("content")).toBeVisible();

		expect(container).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});
});
