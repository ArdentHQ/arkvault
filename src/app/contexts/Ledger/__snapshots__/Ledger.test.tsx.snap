// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Ledger Provider > should render ledger provider 1`] = `
<DocumentFragment>
  <div
    data-testid="content"
  />
</DocumentFragment>
`;
